src/contracts/middleware/IndexRegistry.sol

```diff
@@ -0,0 +1,206 @@
+    }
+
+
+    function _updateTotalOperatorHistory(uint8 quorumNumber, uint32 numOperators) internal {
+
+        uint256 totalOperatorsHistoryLength = totalOperatorsHistory[quorumNumber].length;
+
+        //if there is a prior entry, update its "toBlockNumber"
+        if (totalOperatorsHistoryLength > 0) {
+            totalOperatorsHistory[quorumNumber][totalOperatorsHistoryLength - 1].toBlockNumber = uint32(block.number);
+        }
+
+        OperatorIndex memory totalOperatorUpdate;
+        // In the case of totalOperatorsHistory, the index parameter is the number of operators in the quorum
+        totalOperatorUpdate.index = numOperators;
+        totalOperatorsHistory[quorumNumber].push(totalOperatorUpdate);
+    }
+
+    /// 
+    /// @param operatorId operatorId of the operator to update
+    /// @param quorumNumber quorumNumber of the operator to update
+    /// @param index the latest index of that operator in the list of operators registered for this quorum
+    function _updateOperatorIdToIndexHistory(bytes32 operatorId, uint8 quorumNumber, uint32 index) internal {
+        uint256 operatorIdToIndexHistoryLength = operatorIdToIndexHistory[operatorId][quorumNumber].length;
+        //if there is a prior entry for the operator, set the previous entry's toBlocknumber
+        if (operatorIdToIndexHistoryLength > 0) {
+            operatorIdToIndexHistory[operatorId][quorumNumber][operatorIdToIndexHistoryLength - 1].toBlockNumber = uint32(block.number);
+        }
+        OperatorIndex memory latestOperatorIndex;
+        latestOperatorIndex.index = index;
+        operatorIdToIndexHistory[operatorId][quorumNumber].push(latestOperatorIndex);
+
+        emit QuorumIndexUpdate(operatorId, quorumNumber, index);
+    }
+
+    /// @notice when we remove an operator from a quorum, we simply update the operator's index history
+    ///         as well as any operatorIds we have to swap
+    /// @param quorumNumber quorum number of the operator to remove
+    /// @param indexToRemove index of the operator to remove
+    function _processOperatorRemoval(bytes32 operatorId, uint8 quorumNumber, uint32 indexToRemove, bytes32 operatorIdToSwap) internal {   
+        uint32 operatorIdToSwapIndex = operatorIdToIndexHistory[operatorIdToSwap][quorumNumber][operatorIdToIndexHistory[operatorIdToSwap][quorumNumber].length - 1].index;
+        require(totalOperatorsHistory[quorumNumber][totalOperatorsHistory[quorumNumber].length - 1].index - 1 == operatorIdToSwapIndex, "IndexRegistry._processOperatorRemoval: operatorIdToSwap is not the last operator in the quorum");
+
+        // if the operator is not the last in the list, we must swap the last operator into their positon
+        if(operatorId != operatorIdToSwap){
+            //update the swapped operator's operatorIdToIndexHistory list with a new entry, as their index has now changed
+            _updateOperatorIdToIndexHistory(operatorIdToSwap, quorumNumber, indexToRemove);
```

Comment:
the naming of `indexToRemove` is throwing me a bit here. really we are removing the last index and swapping the `operatorIdToSwap` into the index of the `operatorId`, as this `operatorId` is the ID of the operator being removed.
Also noting that `operatorIdToSwap` must match the last operator in the set of operators for the `quorumNumber`.
IMO the documentation could be better here. This is an important piece of functionality that we need to make sure is clear + correct.

Reply:
does this clear thing up? https://github.com/Layr-Labs/eigenlayer-contracts/pull/76/commits/b03c92cfbb9bfd3366ad0bb1c11c20d3cedf29c1